### PR TITLE
Next working time

### PR DESCRIPTION
**Pull request by james2m**
Added a next_working_time method to allow iterating through working times. Documented and tested.

### CODE REVIEW

**Comment by Intrepidd** [comment]
Hi James,

Thanks for your 2 contributions, I'll try to have a look over the week-end.

With your new methods the code is getting quite redundant, I'll see if I find a way to refactor it a bit to avoid doing the same patterns in each method.

**Comment by james2m** [comment]
Great, thanks, and thanks for making it easy code to work on. I was planning on a couple more pull requests next week too. One to create an enumerable of upcoming hours and another to create an enumerable of upcoming windows within those hours.

File: lib/working_hours/computation.rb [hunk 1]
```diff
@@ -85,6 +85,32 @@ def advance_to_working_time time, config: nil
...
+      config ||= wh_config
+      time = in_config_zone(time, config: config).round
+      loop do
+        # skip holidays and weekends
+        while not working_day?(time, config: config)
+          time = (time + 1.day).beginning_of_day
+        end
+        # find next working range after time
+        time_in_day = time.seconds_since_midnight
+        time = time.beginning_of_day
```
Comment (jarthod) [inline lib/working_hours/computation.rb]: This also won't work if there are several periods in the same day :)

**Comment by jarthod** [comment]
@james2m Thanks for these contributions, indeed there are some redundancy, we can probably refactor this to avoid copying the same code over and over.

I added a comment on your code which doesn't work when there are several working periods in one day.

File: lib/working_hours/computation.rb (hunk 1 above)
Comment (james2m) [inline lib/working_hours/computation.rb]: I thought I got that covered in a test below (line 176), or am I missing something? If you can give me a pointer I'll get it fixed.

**Comment by james2m** [comment]
@jarthod Thanks for the feedback, I'll try and nail the redundancy, I was finding subtle variations between them so I wasn't sure to refactor or not.

Re- the multiple slots, I thought I had that covered in the test suite for both, did I miss something?

Thanks.

File: lib/working_hours/computation.rb (hunk 1 above)
Comment (jarthod) [inline lib/working_hours/computation.rb]: Yep, that's correct, thanks.

**Comment by Intrepidd** [comment]
Maybe I'm misunderstanding the goal of this method but can't you use ``advance_to_closing_time`` from #21 in addition with ``advance_to_working_time`` to achieve what you want ?

**Comment by Intrepidd** [comment]
I made all your specs pass with this implementation : 

```ruby
def next_working_time(time, config: nil)
  time = advance_to_closing_time(time, config: config) if in_working_hours?(time, config: config)
  advance_to_working_time(time, config: config)
end
```

Not sure what the implications regarding performance but this can be a good solution regarding the code redundancy, when I tried merging your 2 branches git got crazy because the code looked alike a lot :)

**Comment by james2m** [comment]
Oh, sorry about that. I should have sent you a pull from my master with both pre-merged. Just thought it best to send separately as you may have wanted to reject one or other.

**Comment by jarthod** [comment]
What about @Intrepidd's implementation suggestion ? i'd much prefer a two line version rather than a copy/paste ;)

**Comment by james2m** [comment]
Sorry for the slow reply I’m in Asia and was travelling over the weekend. 

Yes, it seems perfect. I actually implemented them in the opposite order and didn’t notice that I could re-factor the first after implementing the second. 

I’ll update and send a pull request for tomorrow morning with the additional tests too.

**Comment by jarthod** [comment]
That would be perfect ;)

**Comment by james2m** [comment]
Closing as I've sent a new pull request with the suggested refactor and tests for midnight.

**Pull request closed without merging**